synapse/storage/databases/main/events_worker.py

```diff
@@ -2200,3 +2200,58 @@ def _get_partial_state_events_batch_txn(
             (room_id,),
         )
         return [row[0] for row in txn]
+
+    def mark_event_rejected_txn(
+        self,
+        txn: LoggingTransaction,
+        event_id: str,
+        rejection_reason: Optional[str],
+    ) -> None:
+        """Mark an event that was previously accepted as rejected, or vice versa
+
+        This can happen, for example, when resyncing state during a faster join.
+
+        Args:
+            txn:
+            event_id: ID of event to update
+            rejection_reason: reason it has been rejected, or None if it is now accepted
+        """
+        if rejection_reason is None:
+            logger.info(
+                "Marking previously-processed event %s as accepted",
+                event_id,
+            )
+            self.db_pool.simple_delete_txn(
+                txn,
+                "rejections",
+                keyvalues={"event_id": event_id},
+            )
+        else:
+            logger.info(
+                "Marking previously-processed event %s as rejected(%s)",
+                event_id,
+                rejection_reason,
+            )
+            self.db_pool.simple_upsert_txn(
+                txn,
+                table="rejections",
+                keyvalues={"event_id": event_id},
+                values={
+                    "reason": rejection_reason,
+                    "last_check": self._clock.time_msec(),
+                },
+            )
+        self.db_pool.simple_update_txn(
+            txn,
+            table="events",
+            keyvalues={"event_id": event_id},
+            updatevalues={"rejection_reason": rejection_reason},
+        )
+
+        self.invalidate_get_event_cache_after_txn(txn, event_id)
+
+        # TODO(faster_joins): invalidate the cache on workers. Ideally we'd just
+        #   call '_send_invalidation_to_replication', but we actually need the other
+        #   end to call _invalidate_local_event_cache() rather than (just)
+        #   _get_event_cache.invalidate().
+        #   https://github.com/matrix-org/synapse/issues/12994
```

Comment:
I'm guessing we're avoiding triggering `_invalidate_caches_for_event` on workers because it invalidates too much?

Reply:
I don't think I'd thought about that really. It seems a reasonable solution.

Reply:
```suggestion
        # TODO(faster_joins): invalidate the cache on workers. Ideally we'd just
        #   call '_send_invalidation_to_replication', but we actually need the other
        #   end to call _invalidate_local_event_cache() rather than (just)
        #   _get_event_cache.invalidate().
        #
        #   One solution might be to (somehow) get the workers to call 
        #   _invalidate_caches_for_event() (though that will invalidate more than
        #   strictly necessary).
        #
        #   https://github.com/matrix-org/synapse/issues/12994
```